crates/futures/src/atomics.rs

```diff
@@ -0,0 +1,350 @@
+        // notification was received that the future is ready to make progress.
+        // If seen, however, it probably means that the future is also currently
+        // being polled.
+        Notified,
+
+        // The future is blocked, waiting for something to happen. Stored here
+        // is a self-reference to the future itself so we can pull it out in
+        // `Notify` and continue polling.
+        //
+        // Note that the self-reference here is an Arc-cycle that will leak
+        // memory unless the future completes, but currently that should be ok
+        // as we'll have to stick around anyway while the future is executing!
+        //
+        // This state is removed as soon as a notification comes in, so the leak
+        // should only be "temporary"
+        Waiting(Arc<Package>),
+    }
+
+    struct Waker {
+        value: AtomicI32,
+        notified: AtomicBool,
+    };
+
+    impl Default for Waker {
+        fn default() -> Self {
+            Waker {
+                value: AtomicI32::new(0),
+                notified: AtomicBool::new(false),
+            }
+        }
+    }
+
+    impl Notify for Waker {
+        fn notify(&self, _id: usize) {
+            console_log!("Waker notify");
+            if !self.notified.swap(true, Ordering::SeqCst) {
+                console_log!("Waker, inside if");
+                let _ = unsafe {
+                    core::arch::wasm32::atomic_notify(
+                        &self.value as *const AtomicI32 as *mut i32,
+                        0,
+                    )
+                };
+            }
+        }
+    }
+
+    fn poll_again(package: Arc<Package>) {
+        console_log!("poll_again called");
+        let me = match package.notified.replace(State::Notified) {
+            // we need to schedule polling to resume, so keep going
+            State::Waiting(me) => {
+                console_log!("poll_again Waiting");
+                me
+            }
+
+            // we were already notified, and were just notified again;
+            // having now coalesced the notifications we return as it's
+            // still someone else's job to process this
+            State::Notified => {
+                console_log!("poll_again Notified");
+                return;
+            }
+
+            // the future was previously being polled, and we've just
+            // switched it to the "you're notified" state. We don't have
+            // access to the future as it's being polled, so the future
+            // polling process later sees this notification and will
+            // continue polling. For us, though, there's nothing else to do,
+            // so we bail out.
+            // later see
+            State::Polling => {
+                console_log!("poll_again Polling");
+                return;
+            }
+        };
+
+        let value_location = &package.waker.value as *const AtomicI32 as u32 / 4;
```

Comment:
@alexcrichton here is another place where I'd like to simplify things, is there a better way of obtaining a pointer as u32 number?

Reply:
I actually think it's best to simply cahnge `wait_async` to take `&AtomicI32`, and then you can do different things internally there.